### PR TITLE
[emval] Reuse method calling optimisation in regular calls

### PR DESCRIPTION
**Pull request by RReverser**
Method, function and constructor calls are very similar in nature but used different code paths: method calls have an optimisation where an optimised caller trampoline is generated once and reused for each function signature, whereas the other two serialized arguments and types every time.

There is no reason for them to diverge, so I made function calls and constructors reuse the same optimisation.

Results on a silly microbenchmark running a side-effect-ful JS function with 100 integer args in a loop:

#### with dynamic execution:

as method: 224.504ns
as function: 5367.9ns

#### without dynamic execution:

as method: 845.044ns
as function: 5350.96ns

#### with dynamic execution:

as method: 262.74ns
as function: 246.593ns

#### without dynamic execution:

as method: 859.534ns
as function: 823.498ns

In real world the difference is unlikely to be perceptible, but at the very least this makes the code more consistent and deletes a bunch of duplication.

### CODE REVIEW

**Comment by RReverser** [comment]
FWIW I'm happy to go in the opposite direction too and remove the optimisation if we decide that complexity of maintaining it is not worth it given that speed is measured in microseconds even for complex functions.

I just want those callers to converge on single codepath before I make more PRs tweaking that shared codepath.

File: src/embind/emval.js [hunk 1]
```diff
@@ -381,67 +307,83 @@ var LibraryEmVal = {
...
+        argsList += "arg" + i;
+        params.push("argType" + i);
+        args.push(types[i]);
         functionBody +=
         "    var arg" + i + " = argType" + i + ".readValueFromPointer(args" + (offset ? ("+"+offset) : "") + ");\n";
         offset += types[i]['argPackAdvance'];
     }
     functionBody +=
-        "    var rv = handle[name](" + argsList + ");\n";
+    `    var rv = ${asCtor ? 'new func' : 'func.call'}(${argsList});\n`;
```
Comment (brendandahl) [inline src/embind/emval.js:375]: indent

File: src/embind/emval.js [hunk 2]
```diff
@@ -381,67 +307,83 @@ var LibraryEmVal = {
...
-    }
-
     var signatureName = retType.name + "_$" + types.map(t => t.name).join("_") + "$";
     var functionName = makeLegalFunctionName("methodCaller_" + signatureName);
     var functionBody =
-        "return function " + functionName + "(handle, name, destructors, args) {\n";
+        "return function " + functionName + "(obj, func, destructorsRef, args) {\n";
 
     var offset = 0;
+    var argsList = asCtor ? "" : "obj"; // 'obj?, arg0, arg1, arg2, ... , argN'
```
Comment (brendandahl) [inline src/embind/emval.js:362]: Keep argList an array and join at end with a comma.

File: src/embind/emval.js [hunk 3]
```diff
@@ -381,67 +307,83 @@ var LibraryEmVal = {
...
             "    argType" + i + ".deleteObject(arg" + i + ");\n";
         }
     }
     if (!retType.isVoid) {
-        functionBody +=
-        "    return retType.toWireType(destructors, rv);\n";
+      params.push("emval_returnValue");
+      args.push(emval_returnValue);
+      functionBody +=
+      "    return emval_returnValue(retType, destructorsRef, rv);\n";
```
Comment (brendandahl) [inline src/embind/emval.js:386]: Indention was bad before, but lets fix since we're changing the line.

**Review by brendandahl — CHANGES_REQUESTED** [review]
Looks good, minor nits.

File: system/include/emscripten/val.h [hunk 4]
```diff
@@ -595,13 +578,19 @@ class val {
   template<typename WrapperType>
   friend val internal::wrapped_extend(const std::string& , const val& );
 
-  template<typename Implementation, typename... Args>
-  val internalCall(Implementation impl, Args&&... args) const {
+  template<bool AsCtor, typename Ret, typename Implementation, typename... Args>
```
[the remark this answers is not in the document]
Comment (RReverser) [reply system/include/emscripten/val.h:581]: Heh yeah I am actually changing it to enum in lattter commits but thought I'd leave it for a separate PR. Happy to do that here.

File: src/embind/emval.js (hunk 2 above)
Comment (RReverser) [reply src/embind/emval.js:362]: That's not what original code did, presumably for minor perf boost, so I tried to keep it that way. I don't have a strong preference though.

File: system/include/emscripten/val.h [hunk 5]
```diff
@@ -509,16 +480,28 @@ class val {
     return internal::_emval_delete(as_handle(), val_ref(property).as_handle());
   }
 
+  template<typename... Args>
```
[the remark this answers is not in the document]
Comment (RReverser) [reply system/include/emscripten/val.h:483]: Just to keep all usages of `internalCall` together, I found it makes it easier to iterate on it when you can see all places that need to be updated together on one screen.

File: src/embind/emval.js (hunk 1 above)
Comment (RReverser) [reply src/embind/emval.js:375]: Indentation in this function is generally bad - mixed with 4 and 2 spaces in different places.

I guess I can just update all at once.

File: src/embind/emval.js (hunk 2 above)
Comment (RReverser) [reply src/embind/emval.js]: Changed it to array, although that gets a bit out of scope of this PR.

**Comment by RReverser** [comment]
Addressed the feedback.

**Review by brendandahl — APPROVED** [review]

**Pull request merged**